### PR TITLE
Upgrade to Rails 7 and latest JS/CSS Tooling

### PR DESCRIPTION
**Pull request by iHiD**
### Erik
- [x] Use correct host variable (`Exercism.config.website_assets_host`) in the icon/graphical_icon helper files
- [x] Fix MySQL stuff (maybe pair on this?)
- [x] Implement `--watch` flag for `yarn build` (there's various articles on this online)
- [ ] Click around a lot and look for anything that's broken
- [x] Minify CSS in production
- [x] Minify JS in production

### iHiD
- [ ] Fix any media queries that won't work any more

### CODE REVIEW

File: docker/dev/init.sql [hunk 1]
```diff
@@ -1,7 +1,8 @@
+CREATE USER IF NOT EXISTS 'exercism' IDENTIFIED BY 'exercism';
+
 CREATE DATABASE IF NOT EXISTS `exercism_development`;
 ALTER DATABASE `exercism_development` CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
-/* Implicitly create the user with our first DB to avoid errors if they already exist */
-GRANT ALL PRIVILEGES ON `exercism_development`.* TO 'exercism' IDENTIFIED BY 'exercism';
+GRANT ALL PRIVILEGES ON `exercism_development`.* TO 'exercism';
```
Comment (iHiD) [inline docker/dev/init.sql:5]: Why this change? Just not necessary any more?

Comment (ErikSchierboom) [reply docker/dev/init.sql:5]: It didn't work on my dev-env. I was looking into way to fix it

Comment (iHiD) [reply docker/dev/init.sql:5]: Interesting. Cool.

File: Gemfile [hunk 2]
```diff
@@ -1,9 +1,13 @@
-source 'https://rubygems.org'
+source 'http://rubygems.org'
```
Comment (iHiD) [inline Gemfile]: Revert.

File: Procfile.dev [hunk 3]
```diff
@@ -1,5 +1,7 @@
 server: bin/rails server -p 3020
-assets: bin/webpack-dev-server
-anycable: bundle exec anycable
-sidekiq: bundle exec sidekiq
-ws: anycable-go --host=local.exercism.io --port 3334
+# anycable: bundle exec anycable
+# sidekiq: bundle exec sidekiq
+# ws: anycable-go --host=local.exercism.io --port 3334
```
Comment (iHiD) [inline Procfile.dev]: Revert

File: Procfile.docker.dev [hunk 4]
```diff
@@ -1,4 +1,11 @@
 server: bundle exec bin/rails server -b '0.0.0.0'
-assets: WEBPACKER_DEV_SERVER_HOST=0.0.0.0 bundle exec bin/webpack-dev-server
-anycable: RAILS_ENV=development bundle exec anycable --server-command "anycable-go --disconnect_rate=5 --host=0.0.0.0 --port 3334 --redis_url='redis://redis:6379'"
-sidekiq: bundle exec sidekiq
+# anycable: bundle exec anycable
+# sidekiq: bundle exec sidekiq
+# ws: anycable-go --host=local.exercism.io --port 3334
```
Comment (iHiD) [inline Procfile.docker.dev]: Revert

File: Procfile.docker.dev [hunk 5]
```diff
@@ -1,4 +1,11 @@
...
+css: yarn build:css --watch
+js: yarn build --watch
+
+# assets: WEBPACKER_DEV_SERVER_HOST=0.0.0.0 bundle exec bin/webpack-dev-server
+# anycable: RAILS_ENV=development bundle exec anycable --server-command "anycable-go --disconnect_rate=5 --host=0.0.0.0 --port 3334 --redis_url='redis://redis:6379'"
+# sidekiq: bundle exec sidekiq
```
Comment (iHiD) [inline Procfile.docker.dev]: And/or revert

File: app/css/builds/application.css [hunk 6]
```diff
@@ -0,0 +1,5 @@
+/* Entry point for your PostCSS build */
+
+body {
+    background: green;
+}
```
Comment (iHiD) [inline app/css/builds/application.css]: Delete

File: app/javascript/application.js [hunk 7]
```diff
@@ -0,0 +1 @@
+// Entry point for the build script in your package.json
```
Comment (iHiD) [inline app/javascript/application.js]: Remove

File: app/javascript/esbuild.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env node
+const fs = require('fs')
+const svgrPlugin = require('esbuild-plugin-svgr')
```
Comment (iHiD) [inline app/javascript/esbuild.js]: I think this can be removed now.

File: app/views/layouts/application.html.haml [hunk 9]
```diff
@@ -6,13 +6,16 @@
...
-    = javascript_packs_with_chunks_tag('core', 'data-turbo-track': 'reload', 'data-turbo-eval': false)
-    = javascript_packs_with_chunks_tag(*js_packs, 'data-turbo-track': 'reload', 'data-turbo-eval': false)
-    = javascript_packs_with_chunks_tag(*deferred_js_packs,
-      'data-turbo-track': 'reload',
-      'data-turbo-eval': false,
-      defer: true)
+    = stylesheet_link_tag "application", "data-turbo-track": "reload"
+    = stylesheet_link_tag "internal", "data-turbo-track": "reload"
+    = stylesheet_link_tag "erik", "data-turbo-track": "reload"
+    -# = stylesheet_packs_with_chunks_tag(*css_packs, media: 'all', 'data-turbo-track': 'reload', 'data-turbo-eval': false)
```
Comment (iHiD) [inline app/views/layouts/application.html.haml]: Remove this.

File: app/views/layouts/application.html.haml [hunk 10]
```diff
@@ -6,13 +6,16 @@
...
-    = stylesheet_packs_with_chunks_tag(*css_packs, media: 'all', 'data-turbo-track': 'reload', 'data-turbo-eval': false)
-    = javascript_packs_with_chunks_tag('core', 'data-turbo-track': 'reload', 'data-turbo-eval': false)
-    = javascript_packs_with_chunks_tag(*js_packs, 'data-turbo-track': 'reload', 'data-turbo-eval': false)
-    = javascript_packs_with_chunks_tag(*deferred_js_packs,
-      'data-turbo-track': 'reload',
-      'data-turbo-eval': false,
-      defer: true)
+    = stylesheet_link_tag "application", "data-turbo-track": "reload"
+    = stylesheet_link_tag "internal", "data-turbo-track": "reload"
+    = stylesheet_link_tag "erik", "data-turbo-track": "reload"
```
Comment (iHiD) [inline app/views/layouts/application.html.haml]: We need a better name for this.

File: bin/dev [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+hivemind -p 3020 Procfile.dev
```
Comment (iHiD) [inline bin/dev]: Not sure we need/want this.

File: bin/monitor_manifest [hunk 12]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env ruby
```
Comment (iHiD) [inline bin/monitor_manifest]: Underscore this file name.

File: config/application.rb [hunk 13]
```diff
@@ -38,3 +26,26 @@ class Application < Rails::Application
...
+#           puts "files_to_watch: #{files_to_watch}"
+#           Rails.application.config.file_watcher.new([], files_to_watch) do
+#             "clear cache"
+#             clear_cache
+#           end
+#         end
+#       end
+#     end
+#   end
+# end
```
Comment (iHiD) [inline config/application.rb]: Deleted commented code.

File: config/initializers/new_framework_defaults_7_0.rb [hunk 14]
```diff
@@ -0,0 +1,117 @@
+# Be sure to restart your server when you modify this file.
```
Comment (iHiD) [inline config/initializers/new_framework_defaults_7_0.rb]: Delete this file.

File: dev.Dockerfile [hunk 15]
```diff
@@ -1,7 +1,7 @@
 #############
 ## Stage 1 ##
 #############
-FROM ruby:2.6.6 as builder
+FROM ruby:3.0.3-bullseye as builder
```
Comment (iHiD) [inline dev.Dockerfile]: Could you see if there's 3.1 for these please. May as well jump to that if we can.

File: postcss.config.js [hunk 16]
```diff
@@ -1,17 +1,31 @@
...
+//      autoprefixer: {
+//        flexbox: 'no-2009',
+//      },
+//      features: {
+//        'nesting-rules': true,
+//      },
+//      stage: 3,
+//    }),
+//  ],
+//}
```
Comment (iHiD) [inline postcss.config.js]: Delete this commented code.

File: tsconfig.json [hunk 17]
```diff
@@ -13,6 +13,7 @@
     "strict": true,
     "noEmit": true,
     "allowSyntheticDefaultImports": true,
+    "resolveJsonModule": true,
     "types": ["webpack-env", "app/javascript/declarations"]
   },
   "exclude": [
```
Comment (iHiD) [inline tsconfig.json:19]: 1. Does this minify?
2. Does it target the correct version of JS (and/or add polyfill for older browsers)?
3. Anything else we should do? :)

Comment (ErikSchierboom) [reply tsconfig.json:19]: 1. Yes, as it is just like any other bit of JavaScript (as JSON _is_ JavaScript)
2. No polyfill needed, as JSON is just regular JavaScript.
3. Nope.

FYI The reason this has to be enabled explicitly is that they want to discourage people from reading big JSON files in their projects (which is usually better done via an XHR request). In our case, it's fine.

File: bin/monitor_manifest (hunk 12 above)
Comment (ErikSchierboom) [reply bin/monitor_manifest]: Like `_monitor_manifest`?

File: tsconfig.json (hunk 17 above)
Comment (iHiD) [reply tsconfig.json:19]: Ah, I meant the whole block here, sorry. I mean is esbuild minifying, targeting the correct version, etc.

File: app/views/layouts/application.html.haml (hunk 10 above)
Comment (ErikSchierboom) [reply app/views/layouts/application.html.haml]: An even better name? 😉

File: tsconfig.json (hunk 17 above)
Comment (ErikSchierboom) [reply tsconfig.json:19]: I'll check.

File: bin/monitor_manifest (hunk 12 above)
Comment (iHiD) [reply bin/monitor_manifest]: I meant "hypenate" - sigh!

`monitor-manifest`

File: app/views/layouts/application.html.haml (hunk 10 above)
Comment (iHiD) [reply app/views/layouts/application.html.haml]: 😁

**Comment by iHiD** [comment]
Next up is getting the tests green I think.

File: bin/monitor_manifest (hunk 12 above)
Comment (ErikSchierboom) [reply bin/monitor_manifest]: Right!

File: app/javascript/channels/index.ts [hunk 18]
```diff
@@ -1,5 +1,5 @@
 // Load all the channels within this directory and all subdirectories.
 // Channel files must be named *_channel.js.
 
-const channels = require.context('.', true, /_channel\.js$/)
+const channels = require('./**.*')
```
Comment (ErikSchierboom) [inline app/javascript/channels/index.ts:4]: I think this needs to be fixed

File: .github/workflows/tests.yml [hunk 19]
```diff
@@ -404,4 +404,4 @@ jobs:
         env:
           RAILS_ENV: test
           EXERCISM_CI: true
-        run: bundle exec bin/webpack
+        run: bundle exec rails assets:precompile
```
Comment (ErikSchierboom) [inline .github/workflows/tests.yml]: I'm not sure we need this anymore

File: app/javascript/esbuild.js (hunk 8 above)
Comment (iHiD) [reply app/javascript/esbuild.js]: This plugin can probably still be removed btw.

File: app/javascript/esbuild.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        './app/javascript/packs/internal.tsx',
+      ],
+      bundle: true,
+      sourcemap: true,
+      minify: process.env.NODE_ENV === 'production',
+      watch: process.argv.includes('--watch'),
+      outdir: '.built-assets',
+      tsconfig: './tsconfig.json',
+      define: {
+        'process.env.BUGSNAG_API_KEY': '"938ae3d231c5455e5c6597de1b1467af"',
```
Comment (iHiD) [inline app/javascript/esbuild.js:29]: We should move this into `Exercism.secrets` at some point and retrieve it like below.

Comment (ErikSchierboom) [reply app/javascript/esbuild.js:29]: I added a TODO comment

File: db/seeds.rb [hunk 21]
```diff
@@ -249,10 +249,10 @@
 end
 
 
-exercises = Exercise.all.sort_by{rand}[0,3]
-iHiD.authored_exercises += exercises
+authored_exercises = Exercise.all.excluding(iHiD.authored_exercises).sort_by{rand}[0,3]
+iHiD.authored_exercises += authored_exercises
```
Comment (ErikSchierboom) [inline db/seeds.rb:256]: This fixes an intermittent duplicate key exception, which was caused by the track syncing happening before this where you were already authoring some exercises. The `excluding` method is new in Rails 7 BTW: https://blog.saeloun.com/2021/03/08/rails-6-1-adds-excluding-to-active-record-relation.html

File: .github/workflows/tests.yml [hunk 22]
```diff
@@ -404,4 +405,4 @@ jobs:
...
```
Comment (iHiD) [inline .github/workflows/tests.yml]: This should probably be `test:prepare`

File: app/helpers/react_components/react_component.rb [hunk 23]
```diff
@@ -8,13 +8,10 @@ def to_s(id, data, fitted: false, wrapper_class_modifier: nil, style: nil)
       css_classes << "c-react-wrapper-#{id}-#{wrapper_class_modifier}" if wrapper_class_modifier.present?
       css_classes << '--fitted' if fitted
       tag.div(
-        "",
```
Comment (iHiD) [inline app/helpers/react_components/react_component.rb:11]: I'm not sure this `""` should have been deleted.

File: app/javascript/packs/internal.tsx [hunk 24]
```diff
@@ -1,6 +1,6 @@
 import 'easymde/dist/easymde.min.css'
 
-require('channels')
+import Channels from 'channels'
```
Comment (iHiD) [inline app/javascript/packs/internal.tsx]: What does this do? We don't seem to use `Channels` anywhere in here. Did requiring do something different in the past (like init something?)

Comment (ErikSchierboom) [reply app/javascript/packs/internal.tsx]: Possibly. I'll see if removing breaks anything

**Comment by iHiD** [comment]
Needs:
- #2341 merging into this
- Rebase
- To get green.

Then I'll do a dark-deploy tomorrow and a real deploy over the weekend 🦜

**Comment by iHiD** [comment]
I think we probably need to squash and rebase this to get a clean diff for final review.

**Comment by ErikSchierboom** [comment]
@ihid Done

File: app/javascript/config/.manifest.json.d.ts [hunk 25]
```diff
@@ -0,0 +1,6 @@
+interface Manifest {
+  [key: string]: string
+}
+
+declare const manifest: Manifest
+export default manifest
```
Comment (iHiD) [inline app/javascript/config/.manifest.json.d.ts]: Is there a reason this is `.manifest' with a leading `.` still. I feel like we could just make it `manifest.json` now

File: app/javascript/esbuild.js [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+      splitting: true,
+      minify: process.env.NODE_ENV === 'production',
+      watch: process.argv.includes('--watch'),
+      outdir: '.built-assets',
+      tsconfig: './tsconfig.json',
+      target: browserslistToEsbuild(),
+      define: {
+        // TODO: move bugsnag API key into config
+        'process.env.BUGSNAG_API_KEY': '"938ae3d231c5455e5c6597de1b1467af"',
+        'process.env.WEBSITE_ASSETS_HOST': `"${env['website_assets_host']}"`,
```
Comment (iHiD) [inline app/javascript/esbuild.js:30]: Our of interest do we need the nested string and interpolation thing here?

File: app/javascript/esbuild.js [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+      format: 'esm',
+      splitting: true,
+      minify: process.env.NODE_ENV === 'production',
+      watch: process.argv.includes('--watch'),
+      outdir: '.built-assets',
+      tsconfig: './tsconfig.json',
+      target: browserslistToEsbuild(),
+      define: {
+        // TODO: move bugsnag API key into config
+        'process.env.BUGSNAG_API_KEY': '"938ae3d231c5455e5c6597de1b1467af"',
```
Comment (iHiD) [inline app/javascript/esbuild.js:29]: Is this double quoted deliberately?

File: app/models/submission/file.rb [hunk 28]
```diff
@@ -3,7 +3,7 @@ class Submission::File < ApplicationRecord
 
   attr_writer :content
 
-  URI_REGEX = %r{s3://(?<bucket>[a-z0-9-]+)/(?<key>.*)}.freeze
+  URI_REGEX = %r{s3://(?<bucket>[a-z0-9-]+)/(?<key>.*)}
```
Comment (iHiD) [inline app/models/submission/file.rb:6]: Did this get unfrozen for a reason? Are all strings frozen by default now or something?

File: db/migrate/20211218160303_remove_not_null_on_active_storage_blobs_checksum.active_storage.rb [hunk 29]
```diff
@@ -0,0 +1,6 @@
+# This migration comes from active_storage (originally 20211119233751)
+class RemoveNotNullOnActiveStorageBlobsChecksum < ActiveRecord::Migration[7.0]
+  def change
+    change_column_null(:active_storage_blobs, :checksum, true)
+  end
+end
```
Comment (iHiD) [inline db/migrate/20211218160303_remove_not_null_on_active_storage_blobs_checksum.active_storage.rb:10]: I'd like to change this to specific the algorithm and lock modes to avoid locking the db. I think it's this:

```
execute("ALTER TABLE active_storage_blobs MODIFY COLUMN checksum VARCHAR(255) NULL, ALGORITHM=INPLACE, LOCK=NONE")
```

File: db/migrate/20211218160302_create_active_storage_variant_records.active_storage.rb [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+      config = Rails.configuration.generators
+      config.options[config.orm][:primary_key_type] || :primary_key
+    end
+
+    def blobs_primary_key_type
+      pkey_name = connection.primary_key(:active_storage_blobs)
+      pkey_column = connection.columns(:active_storage_blobs).find { |c| c.name == pkey_name }
+      pkey_column.bigint? ? :bigint : pkey_column.type
+    end
+end
```
Comment (iHiD) [inline db/migrate/20211218160302_create_active_storage_variant_records.active_storage.rb]: This file can be removed.

File: db/migrate/20211218160301_add_service_name_to_active_storage_blobs.active_storage.rb [hunk 31]
```diff
@@ -0,0 +1,18 @@
...
+      end
+
+      change_column :active_storage_blobs, :service_name, :string, null: false
+    end
+  end
+
+  def down
+    remove_column :active_storage_blobs, :service_name
+  end
+end
```
Comment (iHiD) [inline db/migrate/20211218160301_add_service_name_to_active_storage_blobs.active_storage.rb]: This file can be removed.

File: app/javascript/config/.manifest.json.d.ts (hunk 25 above)
Comment (ErikSchierboom) [reply app/javascript/config/.manifest.json.d.ts]: We can. I'll update

File: app/javascript/esbuild.js (hunk 26 above)
Comment (ErikSchierboom) [reply app/javascript/esbuild.js:30]: We do, as esbuild will interpret the value as JSON.

Changing the first option to `'process.env.BUGSNAG_API_KEY': '938ae3d231c5455e5c6597de1b1467af'` results in the following error:

```
$ ./app/javascript/esbuild.js
✘ [ERROR] Invalid define value (must be valid JSON syntax or a single identifier): /assets
```

File: app/javascript/esbuild.js (hunk 27 above)
Comment (ErikSchierboom) [reply app/javascript/esbuild.js:29]: Yep. See above

File: app/models/submission/file.rb (hunk 28 above)
Comment (ErikSchierboom) [reply app/models/submission/file.rb:6]: I think this was rubocop. I'll double check. And does `%r` return a string or a regex object?

Comment (ErikSchierboom) [reply app/models/submission/file.rb:6]:
```
app/models/submission/file.rb:6:15: C: [Correctable] Style/RedundantFreeze: Do not freeze immutable objects, as freezing them has no effect.
  URI_REGEX = %r{s3://(?<bucket>[a-z0-9-]+)/(?<key>.*)}.freeze
```

File: test/test_helper.rb [hunk 32]
```diff
@@ -276,6 +289,10 @@ class ActionDispatch::IntegrationTest
   include Devise::Test::IntegrationHelpers
   include TurboAssertionsHelper
 
+  def setup
+    host! URI(Rails.application.routes.default_url_options[:host]).host
+  end
```
Comment (iHiD) [inline test/test_helper.rb:289]: @ErikSchierboom Always call `super` in setup. A load of tests were breaking because the `ActiveSupport::TestCase#setup` method that we rely on (defined further up) wasn't getting called.

File: app/models/submission/file.rb (hunk 28 above)
Comment (kotp) [reply app/models/submission/file.rb:6]: It returns a `Regexp` which is not mutable.

```irb
3.0.3 :001 > %r{\w+}.class
 => Regexp 
```

Comment (iHiD) [reply app/models/submission/file.rb:6]: Thanks.

File: app/javascript/esbuild.js (hunk 26 above)
Comment (iHiD) [reply app/javascript/esbuild.js:30]: Great. Thanks.

File: db/migrate/20211218160303_remove_not_null_on_active_storage_blobs_checksum.active_storage.rb (hunk 29 above)
Comment (iHiD) [reply db/migrate/20211218160303_remove_not_null_on_active_storage_blobs_checksum.active_storage.rb:10]: I've done this.

File: test/test_helper.rb (hunk 32 above)
Comment (ErikSchierboom) [reply test/test_helper.rb:289]: Ah, gotcha

File: bin/monitor-manifest [hunk 33]
```diff
@@ -1,36 +1,13 @@
...
-end
+require 'generate_js_config'
 
 # Ensure that the files are written when starting
-write_files
+GenerateJSConfig.generate!
 
-unless ENV["EXERCISM_CI"]
-  listener = Listen.to(Rails.root / 'app' / 'images') { write_files }
+if Rails.env.development?
```
Comment (ErikSchierboom) [inline bin/monitor-manifest:9]: Lol, good thing I didn't suggest this in https://github.com/exercism/website/pull/2290/commits/4d692740dc5fa4da279deb11ca1dce6a1ab4b271#diff-d4c7ccc36eef0e673446020de7db1e851adc99a30f3a663e8034b05d5e232df8R32 :D

File: docker/rails.Dockerfile [hunk 34]
```diff
@@ -1,44 +1,35 @@
 FROM ruby:3.1.0-bullseye
 
+# Set global env variables
+ENV RAILS_ENV=development
```
Comment (iHiD) [inline docker/rails.Dockerfile]: This needs to be `production` for the website to work.

File: app/javascript/esbuild.js [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+      entryPoints: [
+        './app/javascript/packs/application.tsx',
+        './app/javascript/packs/core.tsx',
+        './app/javascript/packs/internal.tsx',
+        ...(process.env.RAILS_ENV === 'test'
+          ? ['./app/javascript/packs/test.tsx']
+          : []),
+      ],
+      bundle: true,
+      sourcemap: true,
```
Comment (ErikSchierboom) [inline app/javascript/esbuild.js:19]: Do we want sourcemaps in production?

File: app/javascript/esbuild.js [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+    .build({
+      entryPoints: [
+        './app/javascript/packs/application.tsx',
+        './app/javascript/packs/core.tsx',
+        './app/javascript/packs/internal.tsx',
+        ...(process.env.RAILS_ENV === 'test'
+          ? ['./app/javascript/packs/test.tsx']
+          : []),
+      ],
+      bundle: true,
```
Comment (ErikSchierboom) [inline app/javascript/esbuild.js:18]: Is bundling something we want in development?

File: app/javascript/esbuild.js [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+      sourcemap: true,
+      format: 'esm',
+      splitting: true,
```
Comment (ErikSchierboom) [inline app/javascript/esbuild.js:21]: Is splitting something we want in development?

File: app/javascript/esbuild.js (hunk 35 above)
Comment (iHiD) [reply app/javascript/esbuild.js:19]: Probably not?

File: app/javascript/esbuild.js (hunk 36 above)
Comment (iHiD) [reply app/javascript/esbuild.js:18]: Yeah, I think so. Let's keep that consistent.

File: app/javascript/esbuild.js (hunk 35 above)
Comment (ErikSchierboom) [reply app/javascript/esbuild.js:19]: I agree
